### PR TITLE
Add ignore spawn block monster flag

### PR DESCRIPTION
**Pull request by marmichalski**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

**Issues addressed:** #3600 (spawn block, black knight)
**Supersedes:** #2964

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

**Comment by marmichalski** [comment]
There is a "(list here)" in the issue, but @EPuncker was probably drunk creating the issue so it leads nowhere.

**Review by EPuncker — CHANGES_REQUESTED** [review]
other than missing register_monster_type.lua entry, everything alright

**Comment by EPuncker** [comment]
> There is a "(list here)" in the issue, but @EPuncker was probably drunk creating the issue so it leads nowhere.

I'll fix it, but #2964 for reference

File: data/scripts/lib/register_monster_type.lua [hunk 1]
```diff
@@ -98,7 +98,10 @@ registerMonsterType.flags = function(mtype, mask)
 			mtype:isConvinceable(mask.flags.convinceable)
 		end
 		if mask.flags.summonable ~= nil then
-			mtype.isSummonable(mask.flags.summonable)
```
Comment (marmichalski) [inline data/scripts/lib/register_monster_type.lua:101]: 🙄

File: data/monster/monsters/black_knight.xml [hunk 2]
```diff
@@ -7,6 +7,7 @@
 		<flag summonable="0" />
 		<flag attackable="1" />
 		<flag hostile="1" />
+		<flag ignorespawnblock="1" />
```
Comment (marmichalski) [inline data/monster/monsters/black_knight.xml:10]: I'm not sure if we should be changing that behaviour by default.

Comment (marmichalski) [reply data/monster/monsters/black_knight.xml:10]: What I mean is, there might be projects using this black knight file and having many of them placed on map, so now they'll start respawning wether someone is blocking the spawn or not 🤷

Comment (EPuncker) [reply data/monster/monsters/black_knight.xml:10]: perhaps we should just add the functionality and leave the datapack part to user taste

Comment (marmichalski) [reply data/monster/monsters/black_knight.xml]: I'll add the flag to BK, but with value 0 😎 

💩 
🔽 
🚽 

or just simply remove it, someone will just have to find out what flags are supported 🥇

**Review by EPuncker — APPROVED** [review]

**Review by nekiro — DISMISSED** [review]

**Review by DSpeichert — CHANGES_REQUESTED** [review]
I see `ignoresSpawnBlock` and `ignorespawnblock` used interchangeably. Extremely confusing, let's settle on one spelling throughout.

**Comment by EPuncker** [comment]
🤔 by the way can you add the delayed spawn feature together with this PR? #2964

**Comment by marmichalski** [comment]
> 🤔 by the way can you add the delayed spawn feature together with this PR? #2964

No. I can do that in next PR.

**Comment by Erza** [comment]
> I see `ignoresSpawnBlock` and `ignorespawnblock` used interchangeably. Extremely confusing, let's settle on one spelling throughout.

It's just in XML, like every other flag. All flags are lowercase in XML.

XML
```xml
<flag ignorespawnblock="0" />
<flag canpushitems="1" />
<flag canpushcreatures="1" />
<flag targetdistance="1" />
<flag staticattack="90" />
<flag runonhealth="0" />
```

C++
```cpp
bool ignoresSpawnBlock = false;
bool canPushItems = false;
bool canPushCreatures = false;
int32_t targetDistance = 1;
uint32_t staticAttackChance = 95;
int32_t runAwayHealth = 0;
```

**Review by Erza — DISMISSED** [review]

**Comment by DSpeichert** [comment]
`ignoresSpawnBlock` => `ignores Spawn Block`

`ignorespawnblock` => `ignore spawn block`


The missing `s` looks like a typo, and if it's not a typo, it's an unnecessary difference. I know other attributes differ in naming between XML and C++ but there's no reason for them to be different here other than to confuse the reader.

**Comment by marmichalski** [comment]
Contributing to this project is worse than walking on shattered glass. Even when something gets approved, after wave of nonsense/offtopic/useless comments, it gets stalled forever. Truly the most demotivating project I have ever contributed to.

**Comment by joseluis2g** [comment]
yeah because trying to correct a typo is the worst thing you can do

**Comment by marmichalski** [comment]
> yeah because trying to correct a typo is the worst thing you can do

This is not a typo and this is not about this **SINGLE** PR.

**Comment by nekiro** [comment]
> Contributing to this project is worse than walking on shattered glass. Even when something gets approved, after wave of nonsense/offtopic/useless comments, it gets stalled forever. Truly the most demotivating project I have ever contributed to.

Yep and your attitude isn't helping either.

File: data/monster/lua/#example.lua [hunk 3]
```diff
@@ -24,6 +24,7 @@ monster.flags = {
 	hostile = true,
 	challengeable = true,
 	convinceable = false,
+	ignoreSpawnBlock = false,
```
Comment (EPuncker) [inline data/monster/lua/#example.lua:27]:
```suggestion
	ignoresSpawnBlock = false,
```

File: data/monster/monsters/black_knight.xml [hunk 4]
```diff
@@ -7,6 +7,7 @@
 		<flag summonable="0" />
 		<flag attackable="1" />
 		<flag hostile="1" />
+		<flag ignorespawnblock="0" />
```
Comment (EPuncker) [inline data/monster/monsters/black_knight.xml:10]:
```suggestion
		<flag ignoresspawnblock="0" />
```

File: src/monsters.cpp [hunk 5]
```diff
@@ -908,6 +908,8 @@ MonsterType* Monsters::loadMonster(const std::string& file, const std::string& m
 				mType->info.isAttackable = attr.as_bool();
 			} else if (strcasecmp(attrName, "hostile") == 0) {
 				mType->info.isHostile = attr.as_bool();
+			} else if (strcasecmp(attrName, "ignorespawnblock") == 0) {
```
Comment (EPuncker) [inline src/monsters.cpp:911]:
```suggestion
			} else if (strcasecmp(attrName, "ignoresspawnblock") == 0) {
```

**Review by EPuncker — APPROVED** [review]

File: data/scripts/lib/register_monster_type.lua [hunk 6]
```diff
@@ -98,7 +98,10 @@ registerMonsterType.flags = function(mtype, mask)
...
+			mtype:isSummonable(mask.flags.summonable)
+		end
+		if mask.flags.ignoreSpawnBlock ~= nil then
+			mtype:ignoresSpawnBlock(mask.flags.ignoreSpawnBlock)
```
Comment (EPuncker) [inline data/scripts/lib/register_monster_type.lua:104]:
```suggestion
			mtype:ignoreSpawnBlock(mask.flags.ignoreSpawnBlock)
```

File: src/luascript.cpp [hunk 7]
```diff
@@ -2777,6 +2777,7 @@ void LuaScriptInterface::registerFunctions()
 	registerClass("MonsterType", "", LuaScriptInterface::luaMonsterTypeCreate);
 	registerMetaMethod("MonsterType", "__eq", LuaScriptInterface::luaUserdataCompare);
 
+	registerMethod("MonsterType", "ignoresSpawnBlock", LuaScriptInterface::luaMonsterTypeIgnoresSpawnBlock);
```
Comment (EPuncker) [inline src/luascript.cpp:2780]:
```suggestion
	registerMethod("MonsterType", "ignoreSpawnBlock", LuaScriptInterface::luaMonsterTypeIgnoreSpawnBlock);
```

File: src/luascript.cpp [hunk 8]
```diff
@@ -12877,6 +12878,23 @@ int LuaScriptInterface::luaMonsterTypeCreate(lua_State* L)
 	return 1;
 }
 
+int LuaScriptInterface::luaMonsterTypeIgnoresSpawnBlock(lua_State* L)
```
Comment (EPuncker) [inline src/luascript.cpp:12881]:
```suggestion
int LuaScriptInterface::luaMonsterTypeIgnoreSpawnBlock(lua_State* L)
```

File: src/luascript.cpp [hunk 9]
```diff
@@ -12877,6 +12878,23 @@ int LuaScriptInterface::luaMonsterTypeCreate(lua_State* L)
...
+{
+	// get: monsterType:ignoresSpawnBlock() set: monsterType:ignoresSpawnBlock(bool)
```
Comment (EPuncker) [inline src/luascript.cpp:12883]:
```suggestion
	// get: monsterType:ignoreSpawnBlock() set: monsterType:ignoreSpawnBlock(bool)
```

File: src/luascript.cpp [hunk 10]
```diff
@@ -12877,6 +12878,23 @@ int LuaScriptInterface::luaMonsterTypeCreate(lua_State* L)
...
+	MonsterType* monsterType = getUserdata<MonsterType>(L, 1);
+	if (monsterType) {
+		if (lua_gettop(L) == 1) {
+			pushBoolean(L, monsterType->info.ignoresSpawnBlock);
```
Comment (EPuncker) [inline src/luascript.cpp:12887]:
```suggestion
			pushBoolean(L, monsterType->info.ignoreSpawnBlock);
```

File: src/luascript.cpp [hunk 11]
```diff
@@ -12877,6 +12878,23 @@ int LuaScriptInterface::luaMonsterTypeCreate(lua_State* L)
...
+		} else {
+			monsterType->info.ignoresSpawnBlock = getBoolean(L, 2);
```
Comment (EPuncker) [inline src/luascript.cpp:12889]:
```suggestion
			monsterType->info.ignoreSpawnBlock = getBoolean(L, 2);
```

File: src/luascript.h [hunk 12]
```diff
@@ -1276,6 +1276,7 @@ class LuaScriptInterface
 		// MonsterType
 		static int luaMonsterTypeCreate(lua_State* L);
 
+		static int luaMonsterTypeIgnoresSpawnBlock(lua_State* L);
```
Comment (EPuncker) [inline src/luascript.h:1279]:
```suggestion
		static int luaMonsterTypeIgnoreSpawnBlock(lua_State* L);
```

File: src/monsters.cpp [hunk 13]
```diff
@@ -908,6 +908,8 @@ MonsterType* Monsters::loadMonster(const std::string& file, const std::string& m
...
+				mType->info.ignoresSpawnBlock = attr.as_bool();
```
Comment (EPuncker) [inline src/monsters.cpp:912]:
```suggestion
				mType->info.ignoreSpawnBlock = attr.as_bool();
```

File: src/monsters.h [hunk 14]
```diff
@@ -147,6 +147,7 @@ class MonsterType
 		bool canPushItems = false;
 		bool canPushCreatures = false;
 		bool pushable = true;
+		bool ignoresSpawnBlock = false;
```
Comment (EPuncker) [inline src/monsters.h:150]:
```suggestion
		bool ignoreSpawnBlock = false;
```

File: src/spawn.cpp [hunk 15]
```diff
@@ -269,7 +269,7 @@ void Spawn::checkSpawn()
 
 		spawnBlock_t& sb = it.second;
 		if (OTSYS_TIME() >= sb.lastSpawn + sb.interval) {
-			if (findPlayer(sb.pos)) {
+			if (!sb.mType->info.ignoresSpawnBlock && findPlayer(sb.pos)) {
```
Comment (EPuncker) [inline src/spawn.cpp:272]:
```suggestion
			if (!sb.mType->info.ignoreSpawnBlock && findPlayer(sb.pos)) {
```

**Review by EPuncker — DISMISSED** [review]

File: src/spawn.cpp [hunk 16]
```diff
@@ -287,18 +287,42 @@ bool Spawn::isInSpawnZone(const Position& pos)
...
 {
-	if (sb.mTypes.size() == 1) {
-		return spawnMonster(spawnId, std::get<0>(sb.mTypes.front()), sb.pos, sb.direction, startup);
-	}
+	bool isBlocked = !startup && findPlayer(sb.pos);
+	size_t monstersCount = sb.mTypes.size(), blockedMonsters = 0;
+
+	auto spawnFunc = [&](bool roll) {
+		MonsterType *mType;
+		for (std::tuple<MonsterType*, uint16_t> tuple : sb.mTypes) {
```
Comment (nekiro) [inline src/spawn.cpp:295]: tuple? why not pair?

**Review by DSpeichert — APPROVED** [review]

**Pull request merged**